BUG: fix Panel.fillna() ignoring axis parameter (re-submission)

stahlous opened the pull request:
closes #3570 
closes #8251

This is a re-submission of PR #8395 and addresses issue #8251. My apologies for the duplicate PR, but despite rebasing to pydata/pandas master, the original PR would not update with my new commits and I couldn't get rid of the `LooseVersion()` errors in the Travis build.

This PR may need some fleshing out still, but I wanted to submit this to see what the thoughts are on this approach. For this PR I've created a new interpolation mechanism at the block level to implement filling across 3 or more dimensions. This avoids the fiddly problems of trying to implement filling of 3+ dimensions at the frame level. However, it isn't possible with this technique to fill across blocks of different dtypes, although that seems like that should be a rare occurrence. 

Incidentally, this will also address #3570, which is one of the issues referenced in #9862.

stahlous committed:
BUG: fix Panel.fillna() ignoring axis parameter

jreback commented on pandas/core/common.py:
```diff
@@ -1764,6 +1764,35 @@ def interpolate_2d(values, method='pad', axis=0, limit=None, fill_value=None, dt
     return values
 
 
+def interpolate_nd(values, method='pad', axis=0, limit=None, fill_value=None, dtype=None):
```
if you are going to do this, ok, but I think a reorg of all of these little functions is in order. If before this PR (e.g. send another), put all of the interp methods that are in ``core/common.py`` into ``core/missing.py`` would be helpful

stahlous commented:
I merged the previous `interpolate_2d()` and `interpolate_nd()` functions into a single function and then renamed it simply `pad()`. I renamed `interpolated_1d()` to simply `interpolate()`. 

Looking through the code, I see a number of places where the terms "pad", "fill", and "interpolate" are used semi-interchangably and with different meanings. I think it would clear things up if, at least internally, these terms had these fixed definitions:

**fill**: fill in missing values with a static value
**pad**: fill in missing values with the most recent non-missing value
**interpolate**: interpolate in the `scipy.interpolate` sense

I'm not suggesting any changes to the public API, but I would be willing to work on a new PR to try and makes things internally consistent. I don't want to confuse this PR by trying to roll it in here.

In any event, this fix seems to work OK. Not sure if it needs more documentation other than mention of bugs fixed in whatsnew.

jreback commented on pandas/core/missing.py:
```diff
@@ -219,15 +219,38 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
+        if ndim == 3:
+            if axis == 0:
+                for n in range(shape[1]):
+                    values[:,n] = func(values[:,n], axis=1)
```
this looks quite duplicative. 
this type of if/then is so error prone. just compute the axis to use.

jreback commented on pandas/tests/test_panel.py:
```diff
@@ -1454,20 +1454,95 @@ def test_fillna(self):
 
+        # either method or value must be specified
         self.assertRaises(ValueError, self.panel.fillna)
+        # method and value can not both be specified
```
add a blank line between tests.

jreback commented on pandas/tests/test_panel.py:
```diff
@@ -1454,20 +1454,95 @@ def test_fillna(self):
         assert_frame_equal(filled['ItemA'],
                            panel['ItemA'].fillna(method='backfill'))
 
+        # Fill forward.
```
add the issue number here as a comment.

stahlous replied:
The issue number is in `test_fillna_axis_0()`. The tests I added to `test_fillna()` don't address any particular issue. There just just isn't much testing for filling in panels yet.

jreback commented on pandas/tests/test_panel.py:
```diff
@@ -1454,20 +1454,95 @@ def test_fillna(self):
+
+        # Forward fill along axis 0, interpolating values across DataFrames.
+        filled = self.panel.fillna(method='ffill', axis=0)
+        nan_indexes = self.panel['ItemB']['C'].index[
```
usually just a simple ``.loc`` is better to do that all of this chained indexing

jreback commented on pandas/tests/test_panel.py:
```diff
@@ -1454,20 +1454,95 @@ def test_fillna(self):
+        # Fill along axis 0 with limit.
+        filled = self.panel.fillna(method='ffill', axis=0, limit=1)
+        a_nan = self.panel['ItemA']['C'].index[
+            self.panel['ItemA']['C'].apply(np.isnan)]
```
just use ``.isnull``

stahlous commented:
updated

jreback commented on pandas/core/internals.py:
```diff
@@ -910,12 +910,12 @@ def _interpolate_with_fill(self, method='pad', axis=0, inplace=False,
-                                    limit=limit,
-                                    fill_value=fill_value,
-                                    dtype=self.dtype)
+        values = mis.pad(values,
```
make this import as ``missing`` and not ``mis``

jreback commented on pandas/core/missing.py:
```diff
@@ -219,16 +219,38 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
+        slice_axis = 1 if axis == 0 else 0
+        slicer = [slice(None), slice(None)]
+
+        if ndim == 3:
```
shoudln't this be the same?

stahlous replied:
Not sure what you mean. The same as what?

The case of ndim == 3 needs to be handled differently than ndim > 3 since the axes are handled differently. Slicing a `Panel` into `DataFrame`s than slicing a `Panel4d` into `Panel`s or any n-dimensional frame into (n-1)-dimensional frames.

jreback commented on pandas/core/missing.py:
```diff
@@ -219,16 +219,38 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
+
+    if ndim > 2:
+        slice_axis = 1 if axis == 0 else 0
+        slicer = [slice(None), slice(None)]
```
use ``slicer = [slice(None)]*ndim

stahlous replied:
Is there a reason to include axes >1? The slice axis will always be either 0 or 1, so the other slices are always implied to be full slices. Is it a matter of clarity?

jreback commented on pandas/core/missing.py:
```diff
@@ -219,16 +219,38 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
-    ndim = values.ndim
-    if values.ndim == 1:
+    if ndim == 1:
         if axis != 0:  # pragma: no cover
```
This should be handled higher up

stahlous commented:
Implemented most of the suggestions - just need clarification on what is meant by about the ndims being the same.

jreback commented on pandas/core/missing.py:
```diff
@@ -219,20 +219,42 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
+
+        if ndim == 3:
+            axis = 0 if (axis > 1) else 1
+        else:
```
you are changing the meaning of axis here. it is exactly ``[slice(None)]*ndim`` indexed by axis for all types. 
this is equivalent to ``.loc[.....]`` and NOT repeated sectioning. I know that may be confusing, but I said that for your tests are doing incorrect indexing. pls make the changes.

jreback commented:
@stahlous do you want to rebase / update. fyi, you might want to squash as well.

stahlous commented:
I squashed the commits. Sorry for the delay - had other things going on the past couple weeks. Can you take a look and let me know if I've addressed everything?

jreback commented on pandas/core/generic.py:
```diff
@@ -2774,40 +2774,28 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
         # can detect if something was set (e.g. in groupby) (GH9221)
         if axis is None:
-            axis = 0
+            axis = self._stat_axis_name
```
need to document this 'change' in the doc-string

jreback commented on pandas/core/generic.py:
```diff
@@ -2774,40 +2774,28 @@ def fillna(self, value=None, method=None, axis=None, inplace=False,
-            if self._is_mixed_type and axis == 1:
+            if self._is_mixed_type:
+                if (self.ndim > 2) and (axis == 0):
+                    raise NotImplementedError('cannot fill across axis 0 for mixed dtypes')
```
too many if statements here, this should be a bit more general I think

stahlous replied:
The way filling across mixed-dtype blocks for 2D frames is done is by transposing axes. This is actually what I proposed doing in my original PR from last year to extend this filling across higher dimensionality frames, but you rejected that idea. I could fill 2D frames using the same general method I'm proposing here, but that will result in in some backwards incompatible behavior and users that depending on being able to fill across mixed-type blocks for 2D would  no longer be able to do so.

jreback replied:
I don't recall that. my point is that this should be much simpler logic. Too many if statements here (and in the internals).

jreback commented on pandas/core/missing.py:
```diff
@@ -219,20 +219,42 @@ def _interpolate_scipy_wrapper(x, y, new_x, method, fill_value=None,
+        else:
+            axis = axis - 1 if (axis > 0) else 0
+
+        for n in range(values.shape[slice_axis]):
```
same comments as before

jreback commented:
closing, we are going to deprecate ``Panel`` in 0.18.0
